model/src/main/kotlin/utils/PurlUtils.kt

```diff
@@ -92,28 +93,44 @@ internal fun createPurl(
 ): String =
     buildString {
         append("pkg:")
-        append(type)
+        append(type.lowercase())
+        append('/')
 
         if (namespace.isNotEmpty()) {
+            append(namespace.trim('/').split('/').joinToString("/") { it.percentEncode() })
             append('/')
-            append(namespace.percentEncode())
+            append(name.trim('/').percentEncode())
+        } else {
+            append(name.percentEncode())
         }
 
-        append('/')
-        append(name.percentEncode())
+        if (version.isNotEmpty()) {
+            append('@')
 
-        append('@')
-        append(version.percentEncode())
+            // See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#character-encoding which
+            // says "the '#', '?', '@' and ':' characters must NOT be encoded when used as separators".
+            val isChecksum = HashAlgorithm.VERIFIABLE.any { version.startsWith("${it.name.lowercase()}:") }
+            if (isChecksum) append(version) else append(version.percentEncode())
+        }
 
-        qualifiers.onEachIndexed { index, entry ->
+        qualifiers.filterValues { it.isNotEmpty() }.toSortedMap().onEachIndexed { index, entry ->
             if (index == 0) append("?") else append("&")
-            append(entry.key.percentEncode())
+
+            val key = entry.key.lowercase()
+            append(key)
+
             append("=")
-            append(entry.value.percentEncode())
+
+            // See https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairs.
+            if (key in KNOWN_KEYS) append(entry.value) else append(entry.value.percentEncode())
         }
 
         if (subpath.isNotEmpty()) {
-            val value = subpath.split('/').joinToString("/", prefix = "#") { it.percentEncode() }
+            val value = subpath.trim('/').split('/')
+                .filter { it.isNotEmpty() && it != "." && it != ".." }
```

Comment:
This filtering does not work for paths like#. `a/b/c/../d`.
Can it actually happen at all that we get any `.` or `..` inside the input?

Reply:
> This filtering does not work for paths like#. `a/b/c/../d`.

Well, the algorithm description says to "Discard empty, '.' and '..' segments", not to normalize paths. Do you think we should still normalize instead?

> Can it actually happen at all that we get any `.` or `..` inside the input?

Why not? The `subpath` is an arbitrary string after all.

Reply:
> Well, the algorithm description says to "Discard empty, '.' and '..' segments", not to normalize paths. 

My bad, I've written this after reading the `purl parsing algo` by accident instead of the `purl encoding algo`.

> Do you think we should still normalize instead?

Hm, dependens on the input. If we can really get paths like I wrote before, then IMO it would make sense. However, I would not require you to do it. Could be done in future. Maybe a code comment would be worth if you agree. I'm also ok to keep as-is.

Reply:
Discarding an intermediate ".." indeed does not make any sense, so I went for normalizing as it at least doe snot break the test suite.